# TEgenomeSimulator

TEgenomeSimulator was created based on [Matias Rodriguez & Wojciech Maka≈Çowski. Software evaluation for de novo detection of transposons. 2022 Mobil DNA](https://mobilednajournal.biomedcentral.com/articles/10.1186/s13100-022-00266-2). The original scripts were from [denovoTE-eval](https://github.com/IOB-Muenster/denovoTE-eval). Some existing functions had been modified and several new functions had been created for TEgenomeSimulator.\

TEgenomeSimulator allows:
1) the simulation of a random genome with multiple chromosomes that are later inserted with TEs randomly;
2) utilise a user-provided genome containing multiple chromosomes where TE bases had been removed, providing a customised genome canvas for random TE insertion.

## Test
```bash
# Random Genome Mode
ml conda
conda activate TEgenomeSimulator

cd /workspace/cflthc/script/TEgenomeSimulator
prefix=test_random
chridx="./test/input/random_genome_chr_index.txt"
repeat="./test/input/combined_curated_TE_lib_ATOSZM_selected.fasta"
max=5
min=1
outdir="./test/output"
mkdir -p $outdir

python3 ./TEgenomeSimulator.py -M 0 -p $prefix -c $chridx -r $repeat -m $max -n $min -o $outdir
```

```bash
# Custom Genome Mode
ml conda
conda activate TEgenomeSimulator

cd /workspace/cflthc/script/TEgenomeSimulator
prefix=test_custom
genome="./test/input/Donghong.chromosomes.only.fa.nonTE.2chrs"
repeat="./test/input/combined_curated_TE_lib_ATOSZM_selected.fasta"
max=5
min=1
outdir="./test/output"
mkdir -p $outdir

python3 ./TEgenomeSimulator.py -M 1 -p $prefix -g $genome -r $repeat -m $max -n $min -o $outdir
```

## To setup required environment, check the following script:
```bash
/workspace/cflthc/script/KRIP_TE/09_benchmarking/TEgenomeSimulator/00_build_conda_env.sh
```

## How to simulate a random genome with multiple chromosomes and random TE insertion?
- an example of prerequisit files:
    - config_random_genome.yml
    - repeats_test.fa
    - repeats_list_test
- an example of output files:
    - ./result/sim_random_genome/genome_out_repeats.fasta
    - ./result/sim_random_genome/genome_out_repeats.gff
    - ./result/sim_random_genome/genome_out_sequence.fasta
    - ./result/sim_random_genome/genome_out_repeats_nest.fasta
    - ./result/sim_random_genome/genome_out_repeats_nest.gff
    - ./result/sim_random_genome/genome_out_sequence_nest.fasta

```bash
ml conda
conda activate TEgenomeSimulator
TEgenomSimulator=/workspace/cflthc/script/KRIP_TE/09_benchmarking/TEgenomeSimulator
python $TEgenomeSimulator/random_genome_seq_TEs.py
python $TEgenomeSimulator/random_genome_nest_TEs.py
```

## How to simulate a custom genome with multiple chromosomes and random TE insertion?
- an example of prerequisit files:
    - config_custom_genome.yml
    - genome.fa
    - repeats_test.fa
    - repeats_list_test
- an example of output files:
    - ./result/sim_custom_genome/custom_repeat_sequence_out.fasta
    - ./result/sim_custom_genome/custom_repeat_annotation_out.gff
    - ./result/sim_custom_genome/custom_genome_sequence_out.fasta
    - ./result/sim_custom_genome/custom_repeat_sequence_out_nest.fasta
    - ./result/sim_custom_genome/custom_repeat_annotation_out_nest.gff
    - ./result/sim_custom_genome/custom_genome_sequence_out_nest.fasta

```bash
ml conda
conda activate TEgenomeSimulator
TEgenomSimulator=/workspace/cflthc/script/KRIP_TE/09_benchmarking/TEgenomeSimulator
$TEgenomeSimulator/custom_genome_seq_TEs.py
$TEgenomeSimulator/custom_genome_nest_TEs.py
```

## How to prepeare a combined TE library from the curated TE libraries of plant model organisms?
- Plant model organisms used:
    - A. thaliana
    - O. sativa
    - Z. maize
- Check the following scripts from processing the curated TE libraries to creating a TE list table for simulation:
```bash
TEgenomSimulator=/workspace/cflthc/script/KRIP_TE/10_TEgenomeSimulator
$TEgenomeSimulator/01_extract_curated_TE_lib.bash
$TEgenomeSimulator/02_stitch_LTR_INT_AT.bash
$TEgenomeSimulator/02_stitch_LTR_INT_OS.bash
$TEgenomeSimulator/02_stitch_LTR_INT.py
$TEgenomeSimulator/03_combine_curated_TE_lib.bash
$TEgenomeSimulator/04_group_curated_TE_lib.bash
$TEgenomeSimulator/05_prep_sim_TE_lib.bash
$TEgenomeSimulator/05_prep_sim_TE_lib.py
```

## How to prepare the _A. chinensis_ (Donghong) genome for the simulation of random TE insertion?
- Identification of repetitive sequences:** 
    - Multiple TE annotators were used to annotate TEs in Donghong genome. The TE annotator included in this research are as follows:
        - EDTA (v2.1) (see /workspace/cflthc/script/KRIP_TE/09_benchmarking/09.01_RM2.ipynb)
        - RepeatModeller2 (see /workspace/cflthc/script/KRIP_TE/09_benchmarking/09.02_EDTA.ipynb)
        - EarlGrey (see /workspace/cflthc/script/KRIP_TE/09_benchmarking/09.03_EarlGrey.ipynb)
- Combining TE libraries from multiple TE annotator results:
    - The TE libraries generated by the tools mentioned above were combined. The exact duplicated sequences were removed using BBMap's `dedupe.sh`. 
    - It is common to have unique TE sequences with identical headers in the combined TE library. Therefore it's better to rename the duplicated headers.
    - see /workspace/cflthc/script/KRIP_TE/09_benchmarking/09.05_RepeatMasker.ipynb
- Masking the repetitive sequences:
    - The combined TE library mentioned in the previous step was imported into RepeatMasker with the option `-x` enabled to convert repeat bases into 'X's
    - The 'X's were then removed simply using `sed` 
    - see /workspace/cflthc/script/KRIP_TE/09_benchmarking/09.05_RepeatMasker.ipynb
- Or Check the following:
```bash
TEgenomSimulator=/workspace/cflthc/script/KRIP_TE/09_benchmarking/TEgenomeSimulator
cat 06_prep_nonTE_genome.txt
```

## How to use TEgenomeSimulator to create random TE insertions in _A. chinensis_ (Donghong) genome where repetitive sequence has been removed?
- Check if the prerequisit files existed:
    - config_custom_genome_DH.yml
    - Donghong.chromosomes.only.fa.nonTE
    - combined_curated_TE_lib_ATOSZM_selected.fasta
    - TElib_sim_list.table
- Modify `custom_genome_seq_TEs.py` to `custom_genome_seq_TEs_DH.py`
- Modify `custom_genome_nest_TEs.py` to `custom_genome_nest_TEs_DH.py`
- an example of output files:
    - ./result/
    
```bash
ml conda
conda activate TEgenomeSimulator
TEgenomeSimulator=/workspace/cflthc/script/KRIP_TE/09_benchmarking/TEgenomeSimulator
cd $TEgenomeSimulator
THREADS=1
RAM=10G
TIME="1-00:00:00"
JOB="TEgenSim"
sbatch --cpus-per-task=$THREADS --mem $RAM -t $TIME -J ${JOB} -o ./${JOB}.out -e ./${JOB}.err --wrap "python $TEgenomeSimulator/custom_genome_seq_TEs_DH_5_10.py"
#sbatch --cpus-per-task=$THREADS --mem $RAM -t $TIME -J ${JOB} -o ./${JOB}.out -e ./${JOB}.err --wrap "python $TEgenomeSimulator/custom_genome_nest_TEs_DH_5_10.py"
```

**Note**: output directory has been changed to /workspace/cflthc/scratch/2022_Actinidia_TE/10.01_TEgenomSimulator_output

