---
title: "A summary of the genome simulated by TEgenomeSimulator"
output: 
  pdf_document:
    toc: true # table of content true
    toc_depth: 2  # upto three depths of headings (specified by #, ## and ###)
    number_sections: true  ## if you want number sections at each table header
    highlight: tango
date: "`r format(Sys.time(), '%d %B, %Y')`"
author: "Ting-Hsuan Chen"
---

```{r setup, include=FALSE}
#range <- '5_500'
genomename <- 'tair10_original'
nonTEgenomefa <- "AthaGenome.fa"
curatedTEfa <- "athaTEref_ru1.classified.fasta"
#TEtablelist <- "TElib_sim_list_tair10.table"
#sim_core <- 1
#sim_time <- "00:05:01"
#sim_mem <- "426.00 MB"

SIM <- paste0("/workspace/cflthc/scratch/2022_Actinidia_TE/10.01_TEgenomeSimulator_output/ori_tair10_genome/")
OUT <- paste0(SIM, "report/")
dir.create(file.path(OUT), recursive = TRUE, showWarnings = FALSE)
knitr::opts_knit$set(root.dir=OUT)
```

```{r, load_TEsnoopeR, include=FALSE}
library(ggplot2)
library(dplyr)
library(tidyr)
library(RColorBrewer)
library(repr)
library(reshape2)
library(stringr)
library(ComplexHeatmap)
library(knitr)
source("/workspace/cflthc/script/TEsnoopeR/R/TEsnoopeR.R")
```

```{r, get_time_stamp, include=FALSE}
#sim_time_stamp <- file.info(path = paste0(SIM, genomename, '_repeat_annotation_out_nest.gff')) |> pull(ctime)
```

\newpage
# Simulation information
__Simulated genome: `r gsub("_.*", "", genomename)`__\

```{r, simulation_info, echo=FALSE}
#sim_info <- data.frame(CPU = sim_core,
#                       RAM = sim_mem,
#                       Time = sim_time,
#                       Timestamp_of_completion = sim_time_stamp)
#kable(sim_info, format="markdown")
```

In this simulation, TEgenomeSimulator did the followings:

1. It took the TE library file, __`r curatedTEfa`__, which comprises curated TE family sequences from __*A. thalian*__, __*Z. maize*__ and __*O. sativa*__, and simulates multiple TE copies with sequence variations depending on the parameters specified in the table: __`r TEtablelist`__. In addition to nucleotide substitution and INDEL, it also simulated fragmentation (as a proportion of TE truncated from 5' end), nested insertion (only for Copia and Gypsy), as well as target site duplication. 

2. The simulated TE copies were then randomly inserted into the user-provided TE-depleted genome, __`r nonTEgenomefa`__, where TE sequences had been exhaustively detected by multiple TE annotators (e.g. EDTA, RepeatModeler, and EarlGrey) and removed. The final simulated genome can be utilised for benchmarking TE annotators.

# Loaded files for creating this report
```{r, load_files, echo=FALSE}
#gfai <- 'AthaGenome.fa.fai'
#tfai <- paste0(genomename, '_repeat_sequence_out_nest.fasta.fai')
#gff <- paste0(genomename, '_repeat_annotation_out_nest.gff')
#genomefai <- read.table(paste0(SIM, gfai), header = F, sep = '\t')
#tefai <- read.table(paste0(SIM, tfai), header = F, sep = '\t', comment.char = "")
#tegff <- read.table(paste0(SIM, gff), header = F, sep = '\t', comment.char = "")

sum_file <- read.table('/workspace/cflthc/script/KRIP_TE/10_TEgenomeSimulator/model_species_RM/tair10/AthaGenome.fa.out.summary', header = T, sep = '\t', comment.char = "")
```

```{r, data_sum, include=FALSE}
data.sum <- sum_file |> group_by(superfamily) |> 
  summarise(count = sum(copynumber), bp = sum(total_length), family_count = n()) |> 
  rename(TE = superfamily) |> 
  mutate(TE = gsub("LTR/Copia", "01_LTR/Copia", TE)) |>
  mutate(TE = gsub("LTR/Gypsy", "02_LTR/Gypsy", TE)) |>
  mutate(TE = gsub("LINE/L1", "03_LINE/L1", TE)) |>
  mutate(TE = gsub("LINE/LINE_Unknown", "04_LINE/unknown", TE)) |>
  mutate(TE = gsub("RathE1_cons", "05_SINE/RathE1_cons", TE)) |>
  mutate(TE = gsub("RathE2_cons", "06_SINE/RathE2_cons", TE)) |>
  mutate(TE = gsub("RathE3_cons", "07_SINE/RathE3_cons", TE)) |>
  mutate(TE = gsub("SINE/SINE_Unknown", "08_SINE/unknown", TE)) |>
  mutate(TE = gsub("DNA/DTC", "09_DNA/DTC", TE)) |>
  mutate(TE = gsub("DNA/DTA", "10_DNA/DTA", TE)) |>
  mutate(TE = gsub("DNA/DTM", "11_DNA/DTM", TE)) |>
  mutate(TE = gsub("DNA/DTH", "12_DNA/DTH", TE)) |>
  mutate(TE = gsub("DNA/DTT", "13_DNA/DTT", TE)) |>
  mutate(TE = gsub("DNA", "14_DNA", TE)) |>
  mutate(TE = gsub("RC/Helitron", "15_RC/Helitron", TE)) |>
  arrange(TE)

```

Note:
- TE families missed from the TEgff file:
  - "ATCOPIA31A" "ATCOPIA47"  "ATCOPIA80"  "TA1-2"
  - Each of these TE families only has a single copy in the tair10 genome.
```
copia_fam_ori <- sum_file |> filter(superfamily == "LTR/Copia") |> arrange(family) |> distinct(family)
copia_fam_sim <- tegff |> mutate(supfamily = gsub(".*Classification=", "", V9)) |> mutate(supfamily = gsub(";.*", "",supfamily)) |> mutate(family = gsub("ID=", "", V9)) |> mutate(family = gsub("#.*", "",family)) |> mutate(family = gsub("_TE.*", "",family)) |> filter(supfamily == "LTR/Copia") |> arrange(family) |> distinct(family)
copia_fam_ori[!(copia_fam_ori$family %in% copia_fam_sim$family),]
[1] "ATCOPIA31A" "ATCOPIA47"  "ATCOPIA80"  "TA1-2"
```



# The proportion of the TE/nonTE sequence

```{r, get_stat, echo=FALSE}
# from RepeatMasker .tbl output file
genomesize <- 119146348
maskedsize <- 16728569
```
From RepeatMasker .tbl output file:
- Total genome size: `r format(genomesize, big.mark=",",scientific=FALSE)` bp
- Total Repeat bases: `r format(tesize, big.mark=",",scientific=FALSE)` bp

```{r, fig1, echo=FALSE, out.width="50%"}
plotwidth <- 8
plotheight <- 8
plot <- plot_sim_genome_bp_piechart(genomesize, maskedsize, genomename)
options(repr.plot.width=plotwidth, repr.plot.height=plotheight)
plot
```

```{r, save_fig1, echo=FALSE}
png(filename = paste0(OUT, "001_ori_genome_size_", genomename, ".png"), width = plotwidth, height = plotheight, units = "in", res = 1200)
print(plot)
while (!is.null(dev.list()))  dev.off()
```

High quality image saved in\
`r OUT`

File name: 001_ori_genome_size_`r genomename`.png 

# Breaking down the simulated TEs by superfamily

There are total 16 TE superfamilies included in the curated TE library. The following table shows the number of loci, bases and family of each superfamily, as well as the percentage of loci, bases and family.  

```{r, table1, echo=FALSE}
total_loci <- sum(data.sum$count)
total_bp <- sum(data.sum$bp)
total_fam <- sum(data.sum$family_count)
data_sum <- data.sum
data_sum$total_loci <- rep(total_loci, nrow(data_sum))
data_sum$total_bp <- rep(total_bp, nrow(data_sum))
data_sum <- data_sum %>% 
  mutate(loci_percentage = round(100*(count/total_loci), 2), bp_percentage = round(100*(bp/total_bp), 2), family_percentage = round(100*(family_count/total_fam), 2)) %>%
  rename(TE_superfamily = TE, loci = count) %>%
  select(-total_loci, -total_bp)
total_row <- data.frame(TE_superfamily = "Total", loci = total_loci, bp = total_bp, family_count = total_fam, loci_percentage = 100, bp_percentage = 100, family_percentage = 100)
data_sum <- rbind(data_sum, total_row)
write.table(data_sum, paste0(OUT, "ori_genome_summary_TEsuperfamily_", genomename, ".csv"), col.names = T, row.names = F, sep = ",")
kable(data_sum, format="markdown")
```

Table saved in\
`r OUT`

File name: sim_genome_summary_TEsuperfamily_`r genomename`.csv 

## Total simulated TE loci categorised by TE superfamily

- Total simulated TE loci: `r format(total_loci, big.mark=",",scientific=FALSE)` loci

```{r, fig2, echo=FALSE, out.width="60%"}
plotwidth <- 8
plotheight <- 8
options(repr.plot.width=plotwidth, repr.plot.height=plotheight)
plot <- plot_sim_te_loci_piechart_tair(data.sum, genomename)
plot
```

```{r, save_fig2, echo=FALSE}
png(filename = paste0(OUT, "002_ori_te_loci_", genomename, ".png"), width = plotwidth, height = plotheight, units = "in", res = 1200)
print(plot)
while (!is.null(dev.list()))  dev.off()
```
High quality image saved in\
`r OUT`

File name: 002_ori_te_loci_`r genomename`.png 

## Total simulated TE bases categorised by TE superfamily

- Total simulated TE bases: `r format(total_bp, big.mark=",",scientific=FALSE)` bp

```{r, fig3, echo=FALSE, out.width="60%"}
plotwidth <- 8
plotheight <- 8
options(repr.plot.width=plotwidth, repr.plot.height=plotheight)
plot <- plot_sim_te_bp_piechart_tair(data.sum, genomename)
plot
```

```{r, save_fig3, echo=FALSE}
png(filename = paste0(OUT, "003_ori_te_bp_", genomename, ".png"), width = plotwidth, height = plotheight, units = "in", res = 1200)
print(plot)
while (!is.null(dev.list()))  dev.off()
```

High quality image saved in\
`r OUT`

File name: 003_ori_te_bp_`r genomename`.png 

## Simulated TE family per superfamily

This part only depends on the curated TE library. It shouldn't make any difference between simulations.

- Total simulated TE families: `r total_fam` families

```{r, fig4, echo=FALSE, out.width="60%"}
plotwidth <- 8
plotheight <- 8
options(repr.plot.width=plotwidth, repr.plot.height=plotheight)
plot <- plot_sim_te_fam_piechart_tair(data.sum, genomename)
plot
```

```{r, save_fig4, echo=FALSE}
png(filename = paste0(OUT, "004_ori_te_family_", genomename, ".png"), width = plotwidth, height = plotheight, units = "in", res = 1200)
print(plot)
while (!is.null(dev.list()))  dev.off()
```

High quality image saved in\
`r OUT`

File name: 004_ori_te_family_`r genomename`.png 

```{r, table2, include=FALSE}
#breakdown <- te_breakdown_sim_genome_tair(tegff)
#write.table(breakdown, paste0(OUT, "sim_genome_TE_insertion_info_", genomename, ".csv"), col.names = T, row.names = F, sep = ",")
```

\newpage
# Extracting full information from TE's gff file

Have a look at the full info extracted from TE's gff file (row 101 to 110):

```{r, show_table2, echo=FALSE}
#kable(breakdown[101:110, 1:6], format="latex",booktabs = TRUE)
#kable(breakdown[101:110, 7:9], format = "latex",booktabs = TRUE)
#kable(breakdown[101:110, 10:15], format = "latex",booktabs = TRUE)
```

Full table saved in\
`r OUT`

File name: sim_genome_TE_insertion_info_`r genomename`.csv 

\newpage
# Simulation of nested TE insertions (not run)

```{r, nested_TE, echo=FALSE}
datasum_te <- data.frame(TE = c("01_LTR/Copia", "02_LTR/Gypsy", "03_LINE/L1", 
                                "04_LINE/unknown", "05_SINE/RathE1_cons", "06_SINE/RathE2_cons", "07_SINE/RathE3_cons", "08_SINE/unknown", "09_DNA/DTC", "10_DNA/DTA", 
                                "11_DNA/DTM", "12_DNA/DTH", "13_DNA/DTT", "14_DNA", "15_RC/Helitron", "Unspecified"))
copia <- breakdown |> filter(TE_rename == "01_LTR/Copia")
gypsy <- breakdown |> filter(TE_rename == "02_LTR/Gypsy")
copia_nest <- copia |> filter(grepl("TEn", ID))
gypsy_nest <- gypsy |> filter(grepl("TEn", ID))
copia_cut <- tegff |> mutate(V10=gsub(".*Cut_by=", "Cut_by=", V9)) |> filter(grepl("Cut_by", V10)) |> filter(grepl("Copia", V10)) |> select(-V10)
gypsy_cut <- tegff |> mutate(V10=gsub(".*Cut_by=", "Cut_by=", V9)) |> filter(grepl("Cut_by", V10)) |> filter(grepl("Gypsy", V10)) |> select(-V10)
datasum_copiacut <- te_by_superfamily_sim_genome_tair(copia_cut) 
datasum_copiacut <- full_join(datasum_copiacut, datasum_te, by = join_by(TE)) |> arrange(TE) %>% replace(is.na(.), 0)
datasum_gypsycut <- te_by_superfamily_sim_genome_tair(gypsy_cut) 
datasum_gypsycut <- full_join(datasum_gypsycut, datasum_te, by = join_by(TE)) |> arrange(TE) %>% replace(is.na(.), 0)
```

## Nested and non-nested Copia (nut run)

```{r, fig5, echo=FALSE, out.width="50%"}
plotwidth <- 8
plotheight <- 8
options(repr.plot.width=plotwidth, repr.plot.height=plotheight)
plot <- plot_sim_nested_loci_piechart(nrow(copia), nrow(copia_nest), genomename, "Copia")
plot
```

```{r, save_fig5, echo=FALSE}
png(filename = paste0(OUT, "005_sim_te_nested_copia_", genomename, ".png"), width = plotwidth, height = plotheight, units = "in", res = 1200)
print(plot)
while (!is.null(dev.list()))  dev.off()
```

High quality image saved in\
`r OUT`

File name: 005_sim_te_nested_copia_`r genomename`.png 

## TE loci cut by nested Copia (not run)

```{r, fig6, echo=FALSE, fig.asp = .67, out.width="80%"}
plotwidth <- 12
plotheight <- 8
options(repr.plot.width=plotwidth, repr.plot.height=plotheight)
plot <- plot_sim_cut_loci_barplot_tair(datasum_copiacut, genomename, "Copia")
plot
```

```{r, save_fig6, echo=FALSE}
png(filename = paste0(OUT, "006_sim_te_cutby_nested_copia_", genomename, ".png"), width = plotwidth, height = plotheight, units = "in", res = 1200)
print(plot)
while (!is.null(dev.list()))  dev.off()
```

High quality image saved in\
`r OUT`

File name: 006_sim_te_cutby_nested_copia_`r genomename`.png 

## Nested and non-nested Gypsy (not run)

```{r, fig7, echo=FALSE, out.width="50%"}
plotwidth <- 8
plotheight <- 8
options(repr.plot.width=plotwidth, repr.plot.height=plotheight)
plot <- plot_sim_nested_loci_piechart(nrow(gypsy), nrow(gypsy_nest), genomename, "Gypsy")
plot
```

```{r, save_fig7, echo=FALSE}
png(filename = paste0(OUT, "007_sim_te_nested_gypsy_", genomename, ".png"), width = plotwidth, height = plotheight, units = "in", res = 1200)
print(plot)
while (!is.null(dev.list()))  dev.off()
```

High quality image saved in\
`r OUT`

File name: 007_sim_te_nested_gypsy_`r genomename`.png 

## TE loci cut by nested Gypsy (not run)

```{r, fig8, echo=FALSE, fig.asp = .67, out.width="80%"}
plotwidth <- 12
plotheight <- 8
options(repr.plot.width=plotwidth, repr.plot.height=plotheight)
plot <- plot_sim_cut_loci_barplot(datasum_gypsycut, genomename, "Gypsy")
plot
```

```{r, save_fig8, echo=FALSE}
png(filename = paste0(OUT, "008_sim_te_cutby_nested_gypsy_", genomename, ".png"), width = plotwidth, height = plotheight, units = "in", res = 1200)
print(plot)
while (!is.null(dev.list()))  dev.off()
```
High quality image saved in\
`r OUT`

File name: 008_sim_te_cutby_nested_gypsy_`r genomename`.png 

\newpage
# The distribution of TE loci identity 

How did TEgenomeSimulator simulate sequence identity in this simulation?

1. The distribution of sequence identity of each TE family was defined by the __idn__  and __sd__ values (user provided) stored in __`r TEtablelist`__. 

2. TEgenomeSimulator took the __idn__ as mean identity and __sd__ as standard deviation to create a distrubution, from which a value was sampled as the __simulated identity__ of a TE member. Therefore, __simulated divergence = 1 - simulated identity__

3. Inherited from it's predecessor, [denovoTE-eval](https://github.com/IOB-Muenster/denovoTE-eval), TEgenomSimulator broke down the __simulated divergence__ into __substitution__ and __INDELs__ (i.e. __divergence = substitution % + INDEL %__). 

4. The __INDEL %__ was defined by the __indels__ value from __`r TEtablelist`__. 

```{r, include=FALSE}
RMout <- read.table("/workspace/cflthc/script/KRIP_TE/10_TEgenomeSimulator/model_species_RM/tair10/AthaGenome.fa.out.processed", header = T, sep = "\t", comment.char = "")
RMout <- rename(RMout, Identity = identity, Integrity = length_ratio) |> 
  mutate(TE_rename = gsub("LTR/Copia", "01_LTR/Copia", superfamily)) |>
  mutate(TE_rename = gsub("LTR/Gypsy", "02_LTR/Gypsy", TE_rename)) |>
  mutate(TE_rename = gsub("LINE/L1", "03_LINE/L1", TE_rename)) |>
  mutate(TE_rename = gsub("LINE/LINE_Unknown", "04_LINE/unknown", TE_rename)) |>
  mutate(TE_rename = gsub("RathE1_cons", "05_SINE/RathE1_cons", TE_rename)) |>
  mutate(TE_rename = gsub("RathE2_cons", "06_SINE/RathE2_cons", TE_rename)) |>
  mutate(TE_rename = gsub("RathE3_cons", "07_SINE/RathE3_cons", TE_rename)) |>
  mutate(TE_rename = gsub("SINE/SINE_Unknown", "08_SINE/unknown", TE_rename)) |>
  mutate(TE_rename = gsub("DNA/DTC", "09_DNA/DTC", TE_rename)) |>
  mutate(TE_rename = gsub("DNA/DTA", "10_DNA/DTA", TE_rename)) |>
  mutate(TE_rename = gsub("DNA/DTM", "11_DNA/DTM", TE_rename)) |>
  mutate(TE_rename = gsub("DNA/DTH", "12_DNA/DTH", TE_rename)) |>
  mutate(TE_rename = gsub("DNA/DTT", "13_DNA/DTT", TE_rename)) |>
  mutate(TE_rename = gsub("DNA$", "14_DNA", TE_rename)) |>
  mutate(TE_rename = gsub("RC/Helitron", "15_RC/Helitron", TE_rename)) 

```




```{r, echo=FALSE, fig9, fig.width = 12, fig.asp = .67}
plotwidth <- 12
plotheight <- 8
options(repr.plot.width=plotwidth, repr.plot.height=plotheight)

plot <- plot_sim_identity_by_loci_tair(RMout)
plot
```

```{r, echo=FALSE}
png(filename = paste0(OUT, "009_ori_te_loci_identity_", genomename, ".png"), width = plotwidth, height = plotheight, units = "in", res = 1200)
print(plot)
while (!is.null(dev.list()))  dev.off()
```
High quality image saved in\
`r OUT`

File name: 009_ori_te_loci_identity_`r genomename`.png 


\newpage
# The distribution of TE loci integrity

How did TEgenomeSimulator simulate sequence integrity in this simulation?

1. TEgenomeSimulator considers sequence integrity as __integrity = (1 - (TE locus length / full length))__

2. The length of a TE locus is decided by INDELs and fragmentation.

3. In the fragmentation step, if a TE length is shorter than 500 bp, TEgenomeSimulator would randomly select a value between 70 and 90 as the fraction of the sequence to be removed from 5' end; otherwise the simulator randomly chooses a value between 40 and 90. (This is the same as in [denovoTE-eval](https://github.com/IOB-Muenster/denovoTE-eval)) 

4. The number of fragmented loci was defined by the value of __frag__ in __`r TEtablelist`__. This value was taken as a proportion of total TE loci to undergo fragmentation step.

```{r, fig10, echo=FALSE, fig.width = 12, fig.asp = .67}
#RMout$Integrity <- round(RMout$Integrity, 2)
plotwidth <- 12
plotheight <- 8
options(repr.plot.width=plotwidth, repr.plot.height=plotheight)
blue <- rev(brewer.pal(n = 9, "YlGnBu")) #TIR
bgreen <- rev(brewer.pal(n=9, "PuBuGn")) #MITE
purple <- rev(brewer.pal(n = 9, "Purples")) #LINE
pink <- rev(brewer.pal(n = 9, "RdPu")) #Helitron
orange <- rev(brewer.pal(n = 9, "YlOrBr")) #LTR
brown <- (brewer.pal(n = 11, "BrBG")) #SINE
mycolor <- c(orange[4:5], purple[2:3], brown[2:5], blue[2:7], pink[5], bgreen[2])

plot <- ggplot(RMout, aes(as.numeric(Integrity), fill = TE_rename)) +
  #call geom_histogram with position="dodge" to offset the bars and manual binwidth of 2
  geom_histogram(position = "stack", binwidth = 0.02, color = "white") #+
  #scale_fill_manual(values=mycolor) +
  #ggtitle(paste0("Simulated genome: ", genomename)) +
  #xlab("Integrity") + ylab("TE loci") +
  #theme(title = element_text(size = 12),
  #      axis.text.y = element_text(size = rel(2)),
  #      axis.text.x = element_text(size = rel(2)),
  #      axis.title.x = element_text(size = rel(1.5)),
  #      axis.title.y = element_text(size = rel(1.5)),
  #      legend.title = element_blank(),
  #      legend.text = element_text(size = rel(1.2)),
  #      legend.key.size = unit(0.5, "cm"),
  #      legend.background = element_blank(),
  #      panel.background = element_rect(fill = "white", color = "black"))
plot
```

```{r, fig10, echo=FALSE, fig.width = 12, fig.asp = .67}
RMout2 <- RMout |> filter(Integrity <=1)
plotwidth <- 12
plotheight <- 8
options(repr.plot.width=plotwidth, repr.plot.height=plotheight)

plot <- plot_sim_integrity_by_loci_tair(RMout2)
plot
```

```{r, save_fig10, echo=FALSE}
png(filename = paste0(OUT, "010_ori_te_loci_integrity_", genomename, ".png"), width = plotwidth, height = plotheight, units = "in", res = 1200)
print(plot)
while (!is.null(dev.list()))  dev.off()
```
High quality image saved in\
`r OUT`

File name: 010_ori_te_loci_integrity_`r genomename`.png 

